conn.go

```diff
@@ -36,6 +36,13 @@ type clientConn struct {
 	wg         sync.WaitGroup
 	sync.Mutex                          // protects inflight
 	inflight   map[uint32]chan<- result // outstanding requests
+	errCh      chan error
+}
+
+// Wait blocks until the conn has shut down, and return the error
+// causing the shutdown.
+func (c *clientConn) Wait() error {
```

Comment:
What happens if two goroutines are calling `Wait()` at the same time?

Reply:
The Wait method should be called once. The one will be exited, the another one will be block forever when SFTP connection closed.

Reply:
If we need Wait method to support called multiple times, may be implement it like ssh Wait method.

https://github.com/golang/crypto/blob/master/ssh/mux.go#L106-L113

Reply:
My intent was to provoke you to think about how this will be used, and what contracts need to be documented and what behavior people would expect from the library.

If we write this such that it may be called once and only once, we need to resolve how we would handle it being called more than once:
1. Document that it cannot be used this way, and then point people to the code comment documenting the behavior when it comes in as a support ticket.
2. Return an error in the case where it is called more than once, document that it returns an error in this case, and then wait for the PR to come in, which adds the functionality of handling an arbitrary number of calls.

Meanwhile, as you note, `ssh.Wait` can handle being called multiple times, and if we're clearly mimicking that library, it would be reasonable for them to expect the same behavior here.

Reply:
The ssh.Wait() method does seem to behave more like @wutaizeng's version. If you call it twice the second call will block forever, even if an error was returned by the call. The documentation isn't clear about this behavior, but here's a quicky example that shows it.

https://gist.github.com/eikenb/d3a92a399cd20beb54f1173246f32e2f

Reply:
Sorry, it seems we were talking 2 different Wait calls. I was looking at the public API Session.Wait, while @wutaizeng was pointing to an internal one on the mux struct.

Reply:
I think it’s still basically relevant that `ssh.Session.Wait` blocks forever if called more than once.

But 😕 I find that confusing behavior. There are a lot of cases where one might want to wait on a connection and do something after there is a disconnect, and at least some of those I could imagine someone wanting to do in separate goroutines.

But then as well, `This package does not fall under the stability promise of the Go language itself, so its API may be changed when pressing needs arise.`

Reply:
I prefer the current behavior to ssh.Wait's block forever on additional read behavior and I don't see any value from modeling ssh's behavior here.